compositor_pipeline/src/queue/internal_queue.rs

```diff
@@ -36,29 +35,37 @@ impl InternalQueue {
         self.timestamp_offsets.remove(input_id);
     }
 
-    pub fn enqueue_frame(&mut self, input_id: InputId, mut frame: Frame) -> Result<(), QueueError> {
-        let next_output_buffer_pts = self.get_next_output_buffer_pts();
+    pub fn input_offset(&self, input_id: &InputId) -> Option<&Duration> {
+        self.timestamp_offsets.get(input_id)
+    }
 
-        match self.inputs_queues.get_mut(&input_id) {
-            Some(input_queue) => {
-                let offset = *self
-                    .timestamp_offsets
-                    .entry(input_id)
-                    .or_insert_with(|| next_output_buffer_pts.saturating_sub(frame.pts));
+    pub fn enqueue_frame(
+        &mut self,
+        input_id: InputId,
+        mut frame: Frame,
+        clock_start: Instant,
+    ) -> Result<(), QueueError> {
+        let Some(input_queue) = self.inputs_queues.get_mut(&input_id) else {
+            return Err(QueueError::UnknownInputId(input_id))
+        };
+
+        let offset = *self
+            .timestamp_offsets
+            .entry(input_id)
+            .or_insert_with(|| clock_start.elapsed().saturating_sub(frame.pts));
 
-                frame.pts += offset;
+        // Modify frame pts to be at the time frame where PTS=0 represent clock_start
+        frame.pts += offset;
 
-                input_queue.push(Arc::new(frame));
-                Ok(())
-            }
-            None => Err(QueueError::UnknownInputId(input_id)),
-        }
+        input_queue.push(frame);
+        Ok(())
     }
 
-    /// Pops frames closest to buffer pts.
+    /// Gets frames closest to buffer pts.
+    ///
     /// Implementation assumes that "useless frames" are already dropped
     /// by [`drop_useless_frames`] or [`drop_pad_useless_frames`]
-    pub fn get_frames_batch(&mut self, buffer_pts: Duration) -> FrameSet<InputId> {
+    pub fn get_frames_batch(&self, buffer_pts: Duration) -> FrameSet<InputId> {
```

Comment:
We can simplify this and remove this assumption entirely, but dropping frames right before getting it from input queue. Sth like this:
```rust
    pub fn get_frames_batch(&mut self, buffer_pts: Duration) -> FrameSet<InputId> {
        let mut frames_batch = FrameSet::new(buffer_pts);

        let inputs: Vec<InputId> = self.inputs_queues.keys().cloned().collect();

        for input_id in inputs {
            if let Some(nearest_frame) = self
                .get_nearest_frame_by_input_id(buffer_pts, &input_id)
                .unwrap()
            {
                frames_batch
                    .frames
                    .insert(input_id.clone(), nearest_frame.clone());
            }
        }

        frames_batch
    }

    fn get_nearest_frame_by_input_id(
        &mut self,
        buffer_pts: Duration,
        input_id: &InputId,
    ) -> Result<Option<Frame>, QueueError> {
        self.drop_frames_by_input_id(input_id, buffer_pts)?;
        let input_queue = self
            .inputs_queues
            .get(input_id)
            .ok_or_else(|| QueueError::UnknownInputId(input_id.clone()))?;
        Ok(input_queue.first().cloned())
    }
```